File: Classes/TellTimeViewController.m
//
//  TellTimeViewController.m
//  KidsTimeFun
//
//  Created by Jagmeet Chawla on 4/12/09.
//  Revised by Krishna Narayan on 9/17/09
//		--Updated graphics
//		--Added tell a friend
//		--Changed app name to Kids Learn To Tell Time
//  Copyright 2009-2013 NSC Partners LLC. Copyright 2014 One Step Ahead Apps, LLC. All rights reserved.
//

#import "TellTimeViewController.h"
#import "ClockView.h"
#import "RandomInteger.h"
#import "KidsTimeFunAppState.h"
#import "AudioPlayer.h"

@implementation TellTimeViewController

@synthesize delegate;
@synthesize isRight;
@synthesize activity;
@synthesize activityType;
@synthesize activityLevel;
@synthesize timeOffset;
@synthesize answerIndex;

 // The designated initializer.  Override if you create the controller programmatically and want to perform customization that is not appropriate for viewDidLoad.
- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    if (self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil]) {
        // Custom initialization
    }
    return self;
}

/*
// Implement loadView to create a view hierarchy programmatically, without using a nib.
- (void)loadView {
}
*/


// Implement viewDidLoad to do additional setup after loading the view, typically from a nib.
- (void)viewDidLoad {
	//Build a clock
	clockView = [[ClockView alloc] initWithFrame:clockContainerView.bounds];
	clockView.backgroundColor = [UIColor colorWithRed:0.0f green:0.0f blue:0.0f	alpha:0.0f];
	
	//initialize intervals and range based on difficulty settings
	int timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices;
	switch (self.activityLevel) {
		case kActLevelYellowBelt:
			numberOfChoices = 2;
			timeInterval = 30;
			timeRangeLow = 30;
			timeRangeHigh = 120;
			break;
		case kActLevelGreenBelt:
			numberOfChoices = 3;
			timeInterval = 15;
			timeRangeLow = 15;
			timeRangeHigh = 240;
			break;
		case kActLevelRedBelt:
			numberOfChoices = 4;
			timeInterval = 5;
			timeRangeLow = 5;
			timeRangeHigh = 360;
			break;
		case kActLevelBlackBelt:
			numberOfChoices = 4;
			timeInterval = 1;
			timeRangeLow = 5;
			timeRangeHigh = 719;
			break;
		default:
			break;
	}
	
	//set number of choices on UI
	switch (numberOfChoices) {
		case 2:
			choices4.hidden = YES;
			choices3.hidden = YES;
			choices2.hidden = NO;
			choices = choices2;
			break;
		case 3:
			choices4.hidden = YES;
			choices3.hidden = NO;
			choices2.hidden = YES;
			choices = choices3;
			break;
		case 4:
			choices4.hidden = NO;
			choices3.hidden = YES;
			choices2.hidden = YES;
			choices = choices4;
			break;
		default:
			break;
	}
	
	//Load first/main clock with Random values
	//Hours first
	randomNumber = [[RandomInteger alloc] initWithRange:1 To:12];
	clockView.hours = randomNumber.randomInteger;
	//Then AM/PM
	clockView.PM = ([randomNumber nextRandomIntegerInRange:0 To:1])?YES:NO;
	//rounded to nearest multiple of timeInterval
	clockView.minutes = ([randomNumber nextRandomIntegerInRange:0 To:59] / timeInterval)*timeInterval;
	//No seconds for now
	clockView.seconds = 0;
	clockView.showSeconds = NO;
	//other clock properties
	clockView.showClockAsAnalog = YES;
	clockView.showMinutesOffsetInHoursHand = YES;
	clockView.showAMPM = NO;
	clockView.showDayNight = NO;
	//add clock as a view in the container
	[clockContainerView addSubview:clockView];
	
	//[clockContainerView2 addSubview:clockView2];
	
	//Then ask the question
	int randomTimeOffset = 0, randomHoursOffset, randomMinutesOffset;
	NSMutableString *labelQuestionText = [NSMutableString string];
	switch (self.timeOffset) {
		case 0:
			labelQuestionText = [NSMutableString stringWithFormat:@"What time is it?"];
			[[AudioPlayer getInstance] playAudioFile:@"what_time_is_it"];
			break;
		case -1:
			randomTimeOffset = [randomNumber nextRandomIntegerInRange:timeRangeLow To:timeRangeHigh];
			randomTimeOffset = (randomTimeOffset/timeInterval)*timeInterval;
			randomHoursOffset = randomTimeOffset/60;
			randomMinutesOffset = randomTimeOffset-(randomHoursOffset*60);
			if (randomHoursOffset == 0) { 
				NSString* minutes = [NSString stringWithFormat:@"%i %@",randomMinutesOffset, randomMinutesOffset == 1 ? @"minute" : @"minutes"];
				labelQuestionText = [NSMutableString stringWithFormat:@"What was the time %@ ago?", minutes];
				[[AudioPlayer getInstance] playTellTime:@"what_was_the_time" playHours:NO hours:@"" playMinutes:YES minutes:minutes playAnd:NO playAgo:YES];
			}
			else {
				if (randomMinutesOffset == 0) {
					NSString* hrs = [NSString stringWithFormat:@"%i %@",randomHoursOffset, randomHoursOffset == 1 ? @"hour" : @"hours"];
					labelQuestionText = [NSMutableString stringWithFormat:@"What was the time %@ ago?",hrs];
					[[AudioPlayer getInstance] playTellTime:@"what_was_the_time" playHours:YES hours:hrs playMinutes:NO minutes:@"" playAnd:NO playAgo:YES];
				}
				else {
					NSString* hrs = [NSString stringWithFormat:@"%i %@",randomHoursOffset, randomHoursOffset == 1 ? @"hour" : @"hours"];
					NSString* minutes = [NSString stringWithFormat:@"%i %@",randomMinutesOffset, randomMinutesOffset == 1 ? @"minute" : @"minutes"];
					labelQuestionText = [NSMutableString stringWithFormat:@"What was the time %@ and %@ ago?", hrs, minutes];
					[[AudioPlayer getInstance] playTellTime:@"what_was_the_time" playHours:YES hours:hrs playMinutes:YES minutes:minutes playAnd:YES playAgo:YES];
				}
			}
			break;
		case 1:
			randomTimeOffset = [randomNumber nextRandomIntegerInRange:timeRangeLow To:timeRangeHigh];
			randomTimeOffset = (randomTimeOffset/timeInterval)*timeInterval;
			randomHoursOffset = randomTimeOffset/60;
			randomMinutesOffset = randomTimeOffset-(randomHoursOffset*60);
			if (randomHoursOffset == 0) { 
				NSString* minutes = [NSString stringWithFormat:@"%i %@",randomMinutesOffset, randomMinutesOffset == 1 ? @"minute" : @"minutes"];
				labelQuestionText = [NSMutableString stringWithFormat:@"What will the time be in %@?", minutes];
				[[AudioPlayer getInstance] playTellTime:@"what_will_the_time_be_in" playHours:NO hours:@"" playMinutes:YES minutes:minutes playAnd:NO playAgo:NO];
			}
			else {
				if (randomMinutesOffset == 0) {
					NSString* hrs = [NSString stringWithFormat:@"%i %@",randomHoursOffset, randomHoursOffset == 1 ? @"hour" : @"hours"];
					labelQuestionText = [NSMutableString stringWithFormat:@"What will the time be in %@?", hrs];
					[[AudioPlayer getInstance] playTellTime:@"what_will_the_time_be_in" playHours:YES hours:hrs playMinutes:NO minutes:@"" playAnd:NO playAgo:NO];

				}
				else {
					NSString* minutes = [NSString stringWithFormat:@"%i %@",randomMinutesOffset, randomMinutesOffset == 1 ? @"minute" : @"minutes"];
					NSString* hrs = [NSString stringWithFormat:@"%i %@",randomHoursOffset, randomHoursOffset == 1 ? @"hour" : @"hours"];

					labelQuestionText = [NSMutableString stringWithFormat:@"What will the time be in %@ and %@?",hrs, minutes];
					[[AudioPlayer getInstance] playTellTime:@"what_will_the_time_be_in" playHours:YES hours:hrs playMinutes:YES minutes:minutes playAnd:YES playAgo:NO];

				}
			}
			break;
		default:
			break;
	}
	//replace 1 hours with 1 hour and 1 minutes with 1 minute
	[labelQuestionText replaceOccurrencesOfString:@" 1 hours" withString:@" 1 hour" options:NSCaseInsensitiveSearch range:NSMakeRange(0, labelQuestionText.length)];
	[labelQuestionText replaceOccurrencesOfString:@" 1 minutes" withString:@" 1 minute" options:NSCaseInsensitiveSearch range:NSMakeRange(0, [labelQuestionText length])];
	
	 //show questions
	labelQuestion.text = labelQuestionText;
	[self.view bringSubviewToFront:labelQuestion];
	 
	//calculate answer
	int answerHours, answerMinutes;
	if (self.timeOffset == 0) {
		answerHours = (int)clockView.hours;
		answerMinutes = (int)clockView.minutes;
	}
	else {
		answerMinutes = (int)clockView.hours*60+(int)clockView.minutes;
		answerMinutes = answerMinutes + (self.timeOffset * randomTimeOffset);
		if (answerMinutes < 0) answerMinutes = answerMinutes+720;
		if (answerMinutes > 720) answerMinutes = answerMinutes-720;
		answerHours = answerMinutes/60;
		answerMinutes = answerMinutes-(answerHours*60);
		if (answerHours == 0) answerHours = 12;
	}
	
	//load random answer choices
	int h=0, m=0, prev_h=0, prev_m=0;
	for (int i = 0; i<choices.numberOfSegments; i++) {
		//generate random time
        h = [randomNumber nextRandomIntegerInRange:1 To:12];
        m = ([randomNumber nextRandomIntegerInRange:0 To:59]/timeInterval)*timeInterval;
		//if same as previous or answer, generate new random minute to change it
		int loopLimit=0;
		while ((loopLimit<9)&&
               (((h*60+m) == (prev_h*60+prev_m))||
                ((h*60+m) == (answerHours*60+answerMinutes)))) {
            //generate random answer again
            h = [randomNumber nextRandomIntegerInRange:1 To:12];
            m = ([randomNumber nextRandomIntegerInRange:0 To:59]/timeInterval)*timeInterval;
            ++loopLimit;
		}
		prev_m = m;
		prev_h = h;
		if (((self.activityLevel == kActLevelYellowBelt) || (self.activityLevel == kActLevelGreenBelt)) && (m == 0)) [choices setTitle:[NSString stringWithFormat:@"%i o'clock", h] forSegmentAtIndex:i];
		else [choices setTitle:[NSString stringWithFormat:@"%i:%02i", h, m] forSegmentAtIndex:i];}
	//load answer in a random index, but remember where it is
	randomNumber.rangeLow = 0;
	randomNumber.rangeHigh = (int) choices.numberOfSegments-1;
    
	self.answerIndex = randomNumber.randomInteger;
	if (((self.activityLevel == kActLevelYellowBelt) || (self.activityLevel == kActLevelGreenBelt)) && (answerMinutes == 0)) [choices setTitle:[NSString stringWithFormat:@"%i o'clock", answerHours] forSegmentAtIndex:self.answerIndex];
		else [choices setTitle:[NSString stringWithFormat:@"%i:%02i", answerHours, answerMinutes] forSegmentAtIndex:self.answerIndex];
	//super load
    [super viewDidLoad];
}

/*
// Override to allow orientations other than the default portrait orientation.
- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation {
    // Return YES for supported orientations
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}
*/

- (void)didReceiveMemoryWarning {
	// Releases the view if it doesn't have a superview.
    [super didReceiveMemoryWarning];
	
	// Release any cached data, images, etc that aren't in use.
}

- (void)viewDidUnload {
	// Release any retained subviews of the main view.
	// e.g. self.myOutlet = nil;
}


- (void)dealloc {
	[randomNumber release];
	[clockContainerView release];
	[clockView release];
	if (self.delegate != self) delegate = nil;
	[choices release];
	[rightOrWrong release];
    [super dealloc];
}

#pragma mark -
#pragma mark Custom Methods

- (IBAction) choicesValueChanged {
	choices.enabled = NO;
	choices.hidden = YES;
	if (choices.selectedSegmentIndex == self.answerIndex) {
		isRight = YES;
		rightOrWrong.image = [UIImage imageNamed:@"GoodJob"];
		rightOrWrong2.image = [UIImage imageNamed:@"Right"];
		rightOrWrong2.hidden = NO;
		[[AudioPlayer getInstance] playCorrectWrong:YES];
		[NSTimer scheduledTimerWithTimeInterval:1.5 target:self selector:@selector(rightAnswer) userInfo:NULL repeats:NO];
	}
	else {
		isRight = NO;
		++wrongCounter;
		rightOrWrong2.image = [UIImage imageNamed:@"Wrong"];
		rightOrWrong.image = [UIImage imageNamed:@"TryAgain"];
		[[AudioPlayer getInstance] playCorrectWrong:NO];
		
		if (self.activityType == kActTypeTimed) {
			rightOrWrong2.hidden = YES;
			rightOrWrong.hidden = NO;
		}
		else {
			rightOrWrong2.hidden = YES;
			rightOrWrong.hidden = NO;
		}
		NSLog(@"Wrong Answer");
		[NSTimer scheduledTimerWithTimeInterval:1.5 target:self selector:@selector(wrongAnswer) userInfo:NULL repeats:NO];
	}
}

- (void) rightAnswer {
	if (wrongCounter > 0) {
		isRight = NO;
		wrongCounter = 0;
	}
	if ([delegate conformsToProtocol:@protocol(DismissActivityDelegate)]) [delegate didDismissActivity:self];
}

- (void) wrongAnswer {
	if (self.activityType != kActTypeTimed) {
		rightOrWrong.image = nil;
		rightOrWrong2.image = nil;
		rightOrWrong2.hidden = YES;
		choices.enabled = YES;
		choices.hidden = NO;
		[self.view setNeedsDisplay];
	}
	else {
		if ([delegate conformsToProtocol:@protocol(DismissActivityDelegate)]) [delegate didDismissActivity:self];
	}

}

@end

File: Classes/TellTimeViewController.swift
// Revised by Krishna Narayan on 5/30/26 — Used Claude to migrate to Swift, fix UI Views, remove deprecations, update for iPad, modernize for Apple UI rules.
// Revised by Krishna Narayan on 6/3/26 — Using Claude changed to 1st, 2nd, and 3rd grade levels, belts are earned not selected, added adaptive weak-drilling algorithm to rectify mistakes and build proficiency after initially providing randomized problems for activities
// Copyright 2026 Island Innovation LLC.  All rights reserved.

import UIKit

@objc(TellTimeViewController)
class TellTimeViewController: BaseViewController {

    weak var delegate: DismissActivityDelegate?
    private(set) var isRight = false
    var activity: Int32 = 0
    var activityType: Int32 = 0
    var activityLevel: Int32 = 0
    var gradeLevel: Int32 = 0
    var timeOffset: Int32 = 0
    var answerIndex: Int = 0

    @IBOutlet private weak var clockContainerView: UIView!
    @IBOutlet private weak var choices2: UISegmentedControl!
    @IBOutlet private weak var choices3: UISegmentedControl!
    @IBOutlet private weak var choices4: UISegmentedControl!
    @IBOutlet private weak var rightOrWrong: UIImageView!
    @IBOutlet private weak var rightOrWrong2: UIImageView?
    @IBOutlet private weak var labelQuestion: UILabel!

    private var clockView: ClockView!
    private var randomNumber: RandomInteger!
    private var choices: UISegmentedControl!
    private var wrongCounter = 0
    private var drilledMinute = 0      // minute bucket this question targets (for adaptive recording)
    private var recordedAnswer = false // ensure we log only the first attempt

    override func viewDidLoad() {
        clockView = ClockView(frame: clockContainerView.bounds)
        clockView.backgroundColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0)

        // Difficulty is now driven by the child's grade level (the single setting),
        // not by an earned belt. Grade sets the clock-time increment, the answer-
        // offset range, and how many answer choices to show.
        let timeInterval: Int, timeRangeLow: Int, timeRangeHigh: Int, numberOfChoices: Int
        switch gradeLevel {
        case kGradeFirst:  (timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices) = (30, 30, 120, 2)
        case kGradeSecond: (timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices) = (15, 15, 240, 3)
        case kGradeThird:  (timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices) = (5, 5, 360, 4)
        default:           (timeInterval, timeRangeLow, timeRangeHigh, numberOfChoices) = (30, 30, 120, 2)
        }

        switch numberOfChoices {
        case 2: choices4.isHidden = true;  choices3.isHidden = true;  choices2.isHidden = false; choices = choices2
        case 3: choices4.isHidden = true;  choices3.isHidden = false; choices2.isHidden = true;  choices = choices3
        default:choices4.isHidden = false; choices3.isHidden = true;  choices2.isHidden = true;  choices = choices4
        }
        styleChoices(choices)

        randomNumber = RandomInteger(range: 1, to: 12)
        clockView.hours = Float(randomNumber.randomInteger)
        clockView.PM = randomNumber.nextRandomInteger(inRange: 0, to: 1) != 0
        // Adaptive drilling: after a random warm-up the engine biases the displayed
        // minute toward the values this child reads wrong most; until then (or for
        // variety) it falls back to a random minute at the grade's increment.
        let randomMinute = (randomNumber.nextRandomInteger(inRange: 0, to: 59) / timeInterval) * timeInterval
        drilledMinute = AdaptiveDrill.shared.nextTargetMinute(grade: gradeLevel, activity: activity, interval: timeInterval) ?? randomMinute
        clockView.minutes = Float(drilledMinute)
        clockView.seconds = 0; clockView.showSeconds = false
        clockView.showClockAsAnalog = true; clockView.showMinutesOffsetInHoursHand = true
        clockView.showAMPM = false; clockView.showDayNight = false
        clockContainerView.addSubview(clockView)

        var randomTimeOffset = 0, randomHoursOffset = 0, randomMinutesOffset = 0
        var questionText = ""
        let audio = AudioPlayer.getInstance()
        switch timeOffset {
        case 0:
            questionText = "What time is it?"
            audio.playAudioFile("what_time_is_it")
        case -1:
            randomTimeOffset = (randomNumber.nextRandomInteger(inRange: timeRangeLow, to: timeRangeHigh) / timeInterval) * timeInterval
            randomHoursOffset = randomTimeOffset / 60; randomMinutesOffset = randomTimeOffset - randomHoursOffset * 60
            let (q, hrStr, mnStr) = buildOffsetQuestion(hrs: randomHoursOffset, mins: randomMinutesOffset)
            questionText = "What was the time \(q) ago?"
            audio.playTellTime("what_was_the_time", playHours: randomHoursOffset > 0, hours: hrStr,
                               playMinutes: randomMinutesOffset > 0, minutes: mnStr,
                               playAnd: randomHoursOffset > 0 && randomMinutesOffset > 0, playAgo: true)
        case 1:
            randomTimeOffset = (randomNumber.nextRandomInteger(inRange: timeRangeLow, to: timeRangeHigh) / timeInterval) * timeInterval
            randomHoursOffset = randomTimeOffset / 60; randomMinutesOffset = randomTimeOffset - randomHoursOffset * 60
            let (q, hrStr, mnStr) = buildOffsetQuestion(hrs: randomHoursOffset, mins: randomMinutesOffset)
            questionText = "What will the time be in \(q)?"
            audio.playTellTime("what_will_the_time_be_in", playHours: randomHoursOffset > 0, hours: hrStr,
                               playMinutes: randomMinutesOffset > 0, minutes: mnStr,
                               playAnd: randomHoursOffset > 0 && randomMinutesOffset > 0, playAgo: false)
        default: break
        }
        labelQuestion.text = questionText
        view.bringSubviewToFront(labelQuestion)

        let answerHours: Int, answerMinutes: Int
        if timeOffset == 0 {
            answerHours = Int(clockView.hours); answerMinutes = Int(clockView.minutes)
        } else {
            var total = Int(clockView.hours) * 60 + Int(clockView.minutes) + Int(timeOffset) * randomTimeOffset
            if total < 0 { total += 720 }; if total > 720 { total -= 720 }
            var ah = total / 60; let am = total - ah * 60
            if ah == 0 { ah = 12 }
            answerHours = ah; answerMinutes = am
        }

        var prevH = 0, prevM = 0
        for i in 0..<choices.numberOfSegments {
            var h = 0, m = 0
            repeat {
                h = randomNumber.nextRandomInteger(inRange: 1, to: 12)
                m = (randomNumber.nextRandomInteger(inRange: 0, to: 59) / timeInterval) * timeInterval
            } while (h * 60 + m == prevH * 60 + prevM || h * 60 + m == answerHours * 60 + answerMinutes)
            prevH = h; prevM = m
            choices.setTitle(choiceLabel(h: h, m: m, grade: Int(gradeLevel)), forSegmentAt: i)
        }

        randomNumber.rangeLow = 0; randomNumber.rangeHigh = choices.numberOfSegments - 1
        answerIndex = randomNumber.randomInteger
        choices.setTitle(choiceLabel(h: answerHours, m: answerMinutes, grade: Int(gradeLevel)), forSegmentAt: answerIndex)

        super.viewDidLoad()

        // Move VoiceOver focus to the question once this activity is on screen.
        if let q = labelQuestion {
            DispatchQueue.main.async { UIAccessibility.post(notification: .screenChanged, argument: q) }
        }
    }

    @IBAction func choicesValueChanged() {
        // Log the first attempt for this question's minute bucket so the adaptive
        // engine learns which times this child struggles with.
        if !recordedAnswer {
            recordedAnswer = true
            AdaptiveDrill.shared.record(grade: gradeLevel, activity: activity,
                                        minute: drilledMinute,
                                        correct: choices.selectedSegmentIndex == answerIndex)
        }
        choices.isEnabled = false; choices.isHidden = true
        if choices.selectedSegmentIndex == answerIndex {
            isRight = true
            rightOrWrong.image = UIImage(named: "GoodJob")
            rightOrWrong2?.image = UIImage(named: "Right"); rightOrWrong2?.isHidden = false
            AudioPlayer.getInstance().playCorrectWrong(true)
            ktfAnnounce("Correct!")
            Timer.scheduledTimer(timeInterval: 1.5, target: self, selector: #selector(rightAnswer), userInfo: nil, repeats: false)
        } else {
            isRight = false; wrongCounter += 1
            rightOrWrong2?.image = UIImage(named: "Wrong"); rightOrWrong.image = UIImage(named: "TryAgain")
            AudioPlayer.getInstance().playCorrectWrong(false)
            ktfAnnounce("Not quite, try again.")
            rightOrWrong2?.isHidden = true; rightOrWrong.isHidden = false
            Timer.scheduledTimer(timeInterval: 1.5, target: self, selector: #selector(wrongAnswer), userInfo: nil, repeats: false)
        }
    }

    @objc private func rightAnswer() {
        if wrongCounter > 0 { isRight = false; wrongCounter = 0 }
        delegate?.didDismissActivity(self)
    }

    @objc private func wrongAnswer() {
        if activityType != kActTypeTimed {
            rightOrWrong.image = nil; rightOrWrong2?.image = nil
            rightOrWrong2?.isHidden = true; choices.isEnabled = true; choices.isHidden = false
            view.setNeedsDisplay()
        } else {
            delegate?.didDismissActivity(self)
        }
    }

    private func choiceLabel(h: Int, m: Int, grade: Int) -> String {
        // Younger grades read whole hours as "o'clock"; third grade uses digital time.
        if (grade == kGradeFirst || grade == kGradeSecond) && m == 0 {
            return "\(h) o'clock"
        }
        return String(format: "%i:%02i", h, m)
    }

    private func buildOffsetQuestion(hrs: Int, mins: Int) -> (String, String, String) {
        let hrStr = hrs > 0 ? "\(hrs) \(hrs == 1 ? "hour" : "hours")" : ""
        let mnStr = mins > 0 ? "\(mins) \(mins == 1 ? "minute" : "minutes")" : ""
        if hrs == 0 { return (mnStr, hrStr, mnStr) }
        if mins == 0 { return (hrStr, hrStr, mnStr) }
        return ("\(hrStr) and \(mnStr)", hrStr, mnStr)
    }
}
